### PR TITLE
RFC: Added recursive transpose to `Diagonal`

### PR DESCRIPTION
**Pull request by andyferris**
I mentioned this one at JuliaLang/LinearAlgebra.jl#408. (This has irked me for some time, but until that discussion I had been pretty sceptical of the foundations of block matrices as `Matrix{Matrix}`)

In theory, `Diagonal` might be used to represent a block-diagonal matrix structure. This takes one small step towards supporting this more fully, and arguably fixes a bug (since transpose is supposed to be recursive, but is not in this case).

@stevengj do you think it is reasonable that `Diagonal` is used for block-diagonal matrices?

### CODE REVIEW

File: base/linalg/diagonal.jl [hunk 1]
```diff
@@ -101,8 +101,10 @@ end
 parent(D::Diagonal) = D.diag
 
 ishermitian(D::Diagonal{<:Real}) = true
-ishermitian(D::Diagonal) = isreal(D.diag)
-issymmetric(D::Diagonal) = true
+ishermitian(D::Diagonal{<:Number}) = isreal(D.diag)
+ishermitian(D::Diagonal) = all(ishermitian, D.diag)
```
Comment (andreasnoack) [inline base/linalg/diagonal.jl]: I think definition also works for the previous signature so maybe delete the previous.

**Review by andreasnoack — APPROVED** [review]

File: base/linalg/diagonal.jl [hunk 2]
```diff
@@ -101,8 +101,10 @@ end
...
+issymmetric(D::Diagonal{<:Number}) = true
+issymmetric(D::Diagonal) = all(issymmetric, D.Diag)
```
Comment (andreasnoack) [inline base/linalg/diagonal.jl]: `Diag` -> `diag`

**Review by andreasnoack — CHANGES_REQUESTED** [review]

File: test/linalg/diagonal.jl [hunk 3]
```diff
@@ -304,3 +304,12 @@ end
...
 end
+
+@testset "block diagonal matrices" begin
+    D1 = Diagonal([[1 2; 3 4]])
+    D2 = Diagonal([[1 1+im; 1-im 1]])
+    @test D1.' == Diagonal([[1 3; 2 4]])
+    @test D2' == D2
+    @test issymmetric(D1) == false
+    @test ishermitian(D2) == true
+end
```
Comment (Sacha0) [inline test/linalg/diagonal.jl]: Perhaps also testing with symmetric and hermitian block diagonal matrices would be worthwhile?

File: base/linalg/diagonal.jl (hunk 1 above)
Comment (andyferris) [inline base/linalg/diagonal.jl]: Yes, it works - but presumably line 104 is a no-op while 105 isn't?

File: base/linalg/diagonal.jl (hunk 2 above)
Comment (andyferris) [inline base/linalg/diagonal.jl]: Thanks

File: test/linalg/diagonal.jl (hunk 3 above)
Comment (andyferris) [inline test/linalg/diagonal.jl]: Yep - I'm adding some extra tests

**Comment by andyferris** [comment]
I made fixes and added `expm`, `logm`, `sqrtm`.

File: base/linalg/diagonal.jl [hunk 4]
```diff
@@ -260,8 +262,10 @@ end
...
 conj(D::Diagonal) = Diagonal(conj(D.diag))
-transpose(D::Diagonal) = D
-ctranspose(D::Diagonal) = conj(D)
+transpose(D::Diagonal{<:Number}) = D
+transpose(D::Diagonal) = Diagonal(transpose.(D.diag))
+ctranspose(D::Diagonal{<:Number}) = conj(D)
+ctranspose(D::Diagonal) = Diagonal(ctranspose.(D.diag))
 
 diag(D::Diagonal) = D.diag
 trace(D::Diagonal) = sum(D.diag)
```
Comment (andyferris) [inline base/linalg/diagonal.jl]: `trace`, `det`, etc confuse me. Are these meant to be recursive? Should these only work for elements which have a ring structure, such as square matrices of the same size (and return such a matrix), or should it reflect any general block-diagonal structure (and return an actual scalar - i.e. `T` in `Diagona{Matrix{T}}`)? 

(@stevengj @StefanKarpinski this reflects exactly what I was trying to get to the bottom of in the long "Taking matrix transposes seriously" discussion - any help would be appreciated greatly)

Comment (StefanKarpinski) [inline base/linalg/diagonal.jl]: I'm not sure, @stevengj may have an opinion. My guess would be that you add/multiply diagonal elements, whatever their type. I doubt these should be recursive, but I don't know.

Comment (andyferris) [inline base/linalg/diagonal.jl]: OK, so the recursive `ctranspose` or `adjoint` isn't aimed at trying to model an arbitrary block matrix *per se* (with blocks of possibly differing size), but is designed to (for instance) facilitate a matrix of (compatible) linear operators? 

This make a *lot* more sense to me!

Comment (andreasnoack) [inline base/linalg/diagonal.jl]: I guess these methods can just fail if the blocks are not compatible for a specific operation. For some operations, it makes perfect sense to allow different sizes of the blocks.

Comment (andyferris) [inline base/linalg/diagonal.jl]: Exactly :)

**Comment by andyferris** [comment]
I *think* this should be good to go, now.

**Review by andreasnoack — APPROVED** [review]

**Pull request merged**